From 04c4a5b5de28741b9512e912a2ae8c268963b975 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Tue, 13 Feb 2024 15:18:56 +0100
Subject: [PATCH] phy: stm32: combophy: Add optional rx equalizer setting

Set Rx equalizer peak voltage filter value from
SYSCFG_COMBOPHY_CR4_RX0_EQ bits.

Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Change-Id: Ifcb001b29cb9683882bba2133c0f80fa179d329b
---
 drivers/phy/st/phy-stm32-combophy.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

--- a/drivers/phy/st/phy-stm32-combophy.c
+++ b/drivers/phy/st/phy-stm32-combophy.c
@@ -22,6 +22,7 @@
 
 #define SYSCFG_COMBOPHY_CR1 0x4C00
 #define SYSCFG_COMBOPHY_CR2 0x4C04
+#define SYSCFG_COMBOPHY_CR4 0x4C0C
 #define SYSCFG_COMBOPHY_CR5 0x4C10
 #define SYSCFG_COMBOPHY_SR  0x4C14
 #define SYSCFG_PCIEPRGCR    0x6080
@@ -42,6 +43,9 @@
 #define SYSCFG_COMBOPHY_CR1_REFSSPEN BIT(18)
 #define SYSCFG_COMBOPHY_CR1_SSCEN BIT(19)
 
+/* SYSCFG CR4 */
+#define SYSCFG_COMBOPHY_CR4_RX0_EQ GENMASK(2, 0)
+
 #define MPLLMULT_19_2 (0x02u << 1)
 #define MPLLMULT_20   (0x7Du << 1)
 #define MPLLMULT_24   (0x68u << 1)
@@ -194,6 +198,17 @@ static int stm32_combophy_pll_init(struc
 				   SYSCFG_COMBOPHY_CR1_SSCEN, SYSCFG_COMBOPHY_CR1_SSCEN);
 	}
 
+	if (!of_property_read_u32(combophy->dev->of_node, "st,rx_equalizer", &val)) {
+		dev_dbg(combophy->dev, "Set RX equalizer %u\n", val);
+		if (val > SYSCFG_COMBOPHY_CR4_RX0_EQ) {
+			dev_err(combophy->dev, "Invalid value %u for rx0 equalizer\n", val);
+			return -EINVAL;
+		}
+
+		regmap_update_bits(combophy->regmap, SYSCFG_COMBOPHY_CR4,
+			   SYSCFG_COMBOPHY_CR4_RX0_EQ, val);
+	}
+
 	if (combophy->type == PHY_TYPE_PCIE) {
 		ret = stm32_impedance_tune(combophy);
 		if (ret) {
